key.go

```diff
@@ -0,0 +1,810 @@
+		if err != nil {
+			return err
+		}
+
+		if k.Algorithm != expectedAlg {
+			return fmt.Errorf(
+				"found algorithm %q (expected %q)",
+				k.Algorithm.String(),
+				expectedAlg.String(),
+			)
+		}
+	}
+
+	return nil
+}
+
+type keyalias Key
+
+type marshaledKey struct {
+	keyalias
+
+	// RawAlgorithm contains the raw Algorithm value, this is necessary
+	// because cbor library ignores omitempty on types that implement the
+	// cbor.Marshaler interface.
+	RawAlgorithm cbor.RawMessage `cbor:"3,keyasint,omitempty"`
+
+	// RawKeyParam contains the raw CBOR encoded data for the label -1.
+	// Depending on the KeyType this is used to populate either Curve or K
+	// below.
+	RawKeyParam cbor.RawMessage `cbor:"-1,keyasint,omitempty"`
+}
+
+// MarshalCBOR encodes Key into a COSE_Key object.
+func (k *Key) MarshalCBOR() ([]byte, error) {
+	tmp := marshaledKey{
+		keyalias: keyalias(*k),
+	}
+	var err error
+
+	switch k.KeyType {
+	case KeyTypeSymmetric:
+		if tmp.RawKeyParam, err = encMode.Marshal(k.K); err != nil {
+			return nil, err
+		}
+	case KeyTypeEc2, KeyTypeOkp:
+		if tmp.RawKeyParam, err = encMode.Marshal(k.Curve); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("invalid key type: %q", k.KeyType.String())
+	}
+
+	if k.Algorithm != AlgorithmInvalid {
+		if tmp.RawAlgorithm, err = encMode.Marshal(k.Algorithm); err != nil {
+			return nil, err
+		}
+	}
+
+	return encMode.Marshal(tmp)
+}
+
+// UnmarshalCBOR decodes a COSE_Key object into Key.
+func (k *Key) UnmarshalCBOR(data []byte) error {
+	var tmp marshaledKey
+
+	if err := decMode.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+	*k = Key(tmp.keyalias)
+
+	if tmp.RawAlgorithm != nil {
+		if err := decMode.Unmarshal(tmp.RawAlgorithm, &k.Algorithm); err != nil {
+			return err
+		}
+	}
+
+	switch k.KeyType {
+	case KeyTypeEc2:
+		if tmp.RawKeyParam == nil {
+			return errors.New("missing Curve parameter (required for EC2 key type)")
+		}
+
+		if err := decMode.Unmarshal(tmp.RawKeyParam, &k.Curve); err != nil {
+			return err
+		}
+	case KeyTypeOkp:
+		if tmp.RawKeyParam == nil {
+			return errors.New("missing Curve parameter (required for OKP key type)")
+		}
+
+		if err := decMode.Unmarshal(tmp.RawKeyParam, &k.Curve); err != nil {
+			return err
+		}
+	case KeyTypeSymmetric:
+		if tmp.RawKeyParam == nil {
+			return errors.New("missing K parameter (required for Symmetric key type)")
+		}
+
+		if err := decMode.Unmarshal(tmp.RawKeyParam, &k.K); err != nil {
+			return err
+		}
+	default:
+		// this should not be reachable as KeyType.UnmarshalCBOR would
+		// result in an error during decMode.Unmarshal() above, if the
+		// value in the data doesn't correspond to one of the above
+		// types.
+		return fmt.Errorf("unexpected key type %q", k.KeyType.String())
+	}
+
+	return k.Validate()
+}
+
+// PublicKey returns a crypto.PublicKey generated using Key's parameters.
+func (k *Key) PublicKey() (crypto.PublicKey, error) {
```

Comment:
I was thinking about `Public() crypto.PublicKey` just like the [crypto.Signer](https://pkg.go.dev/crypto#Signer) interface. However, I think we cannot do `Public()` as we need to return errors.

Reply:
I guess that means this should be left as `PublicKey()`?